File: Backend/src/main/java/com/example/freelanceapp/entities/Invoice.java
package com.example.freelanceapp.entities;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "invoices")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Invoice {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "client_id", nullable = false)
    private Client client;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id")
    private Project project;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "quote_id")
    private Quote quote;

    @Column(
        name = "invoice_number",
        nullable = false,
        unique = true,
        length = 50,
        insertable = false,
        updatable = false
    )
    private String invoiceNumber;

    @Column(name = "title", nullable = false, length = 255)
    private String title;

    @Column(name = "status", length = 20)
    private String status = "draft"; // draft, sent, viewed, partial, paid, overdue, cancelled

    @Column(name = "issue_date", nullable = false)
    private LocalDate issueDate;

    @Column(name = "due_date", nullable = false)
    private LocalDate dueDate;

    @Column(name = "paid_date")
    private LocalDate paidDate;

    @Column(name = "payment_terms", length = 100)
    private String paymentTerms; // Net 30, Due on receipt, etc.

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Column(name = "terms", columnDefinition = "TEXT")
    private String terms;

    @Column(name = "subtotal", precision = 10, scale = 2, nullable = false)
    private BigDecimal subtotal = BigDecimal.ZERO;

    @Column(name = "tax_amount", precision = 10, scale = 2)
    private BigDecimal taxAmount = BigDecimal.ZERO;

    @Column(name = "discount_amount", precision = 10, scale = 2)
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @Column(name = "total_amount", precision = 10, scale = 2, nullable = false)
    private BigDecimal totalAmount = BigDecimal.ZERO;

    @Column(name = "amount_paid", precision = 10, scale = 2)
    private BigDecimal amountPaid = BigDecimal.ZERO;

    @Column(name = "balance_due", precision = 10, scale = 2, nullable = false)
    private BigDecimal balanceDue = BigDecimal.ZERO;

    @Column(name = "currency", length = 3)
    private String currency = "USD";

    @Column(name = "sent_at")
    private LocalDateTime sentAt;

    @Column(name = "viewed_at")
    private LocalDateTime viewedAt;

    @Column(name = "pdf_url", columnDefinition = "TEXT")
    private String pdfUrl;

    @Column(name = "public_hash", unique = true, length = 64)
    private String publicHash;

    @Column(name = "payment_link", columnDefinition = "TEXT")
    private String paymentLink; // Stripe/PayPal payment link

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: Frontend/src/types/invoice.ts
export type InvoiceStatus = 'draft' | 'sent' | 'viewed' | 'partial' | 'paid' | 'overdue' | 'cancelled';
export type PaymentMethod = 'stripe' | 'paypal' | 'bank_transfer' | 'cash' | 'check' | 'other';
export type PaymentStatus = 'pending' | 'completed' | 'failed' | 'refunded';

export interface InvoiceItem {
  id: string;
  invoiceId: string;
  description: string;
  quantity: number;
  unitPrice: number;
  taxRate: number;
  discount: number;
  total: number;
  sortOrder: number;
  createdAt: string;
}

export interface InvoicePayment {
  id: string;
  invoiceId: string;
  paymentMethod: PaymentMethod;
  transactionId?: string;
  amount: number;
  currency: string;
  paymentDate: string;
  notes?: string;
  status: PaymentStatus;
  metadata?: any;
  createdAt: string;
}

export interface Invoice {
  id: string;
  userId: string;
  clientId: string;
  projectId?: string;
  quoteId?: string;
  clientName: string;
  clientContactName: string;
  projectName?: string;
  quoteTitle?: string;
  invoiceNumber: string;
  title: string;
  status: InvoiceStatus;
  issueDate: string;
  dueDate: string;
  paidDate?: string;
  paymentTerms?: string;
  notes?: string;
  terms?: string;
  subtotal: number;
  taxAmount: number;
  discountAmount: number;
  totalAmount: number;
  amountPaid: number;
  balanceDue: number;
  currency: string;
  sentAt?: string;
  viewedAt?: string;
  pdfUrl?: string;
  publicHash?: string;
  paymentLink?: string;
  items: InvoiceItem[];
  payments: InvoicePayment[];
  createdAt: string;
  updatedAt: string;
}

export interface InvoiceSummary {
  id: string;
  invoiceNumber: string;
  title: string;
  status: InvoiceStatus;
  clientName: string;
  totalAmount: number;
  amountPaid: number;
  balanceDue: number;
  issueDate: string;
  dueDate: string;
  daysUntilDue?: number;
  isOverdue: boolean;
}

export interface InvoiceAgingReport {
  clientName: string;
  totalBalance: number;
  current: number;
  overdue1_30: number;
  overdue31_60: number;
  overdue61_90: number;
  overdue90_plus: number;
}

export interface CreateInvoiceRequest {
  clientId: string;
  projectId?: string;
  quoteId?: string;
  title: string;
  invoiceNumber?: string;
  issueDate: string;
  dueDate: string;
  paymentTerms?: string;
  notes?: string;
  terms?: string;
  taxAmount?: number;
  discountAmount?: number;
  currency?: string;
  items?: CreateInvoiceItemRequest[];
}

export interface UpdateInvoiceRequest {
  title?: string;
  issueDate?: string;
  dueDate?: string;
  paymentTerms?: string;
  notes?: string;
  terms?: string;
  taxAmount?: number;
  discountAmount?: number;
  status?: InvoiceStatus;
  currency?: string;
}

export interface CreateInvoiceItemRequest {
  description: string;
  quantity: number;
  unitPrice: number;
  taxRate?: number;
  discount?: number;
  sortOrder?: number;
}

export interface UpdateInvoiceItemRequest {
  description?: string;
  quantity?: number;
  unitPrice?: number;
  taxRate?: number;
  discount?: number;
  sortOrder?: number;
}

export interface CreatePaymentRequest {
  paymentMethod: PaymentMethod;
  transactionId?: string;
  amount: number;
  currency?: string;
  paymentDate: string;
  notes?: string;
  status?: PaymentStatus;
}

export interface UpdatePaymentRequest {
  paymentMethod?: PaymentMethod;
  transactionId?: string;
  amount?: number;
  currency?: string;
  paymentDate?: string;
  notes?: string;
  status?: PaymentStatus;
}

export interface InvoicesResponse {
  content: Invoice[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
  first: boolean;
  last: boolean;
  empty: boolean;
}

export interface InvoiceStats {
  totalInvoices: number;
  draftInvoices: number;
  sentInvoices: number;
  partialInvoices: number;
  paidInvoices: number;
  overdueInvoices: number;
  totalInvoicedAmount: number;
  totalAmountPaid: number;
  totalBalanceDue: number;
  averageInvoiceValue: number;
  onTimePaymentRate: number;
  overdueAmount: number;
}